README.txt
SmartFanControl v0.8.6

SmartFanControl - keep it cool - service containing all in one fan and temperature control
=============== = ==== == ==== = ======= ========== === == === === === =========== =======

pigpio is used for IO to rapsberry pi

This program can run as systemd service or as stand-alone program.
As systemd service it starts after pigpiod.

The service is completely configurable via smartfancontrol.xml, located in /etc/smartfancontrol.xml.

	This xml file describes the settings for smartfancontrol.

	Keep item names as they are, otherwise defaults will be used.

	The parameters can be modified, some parameters are only used in a specific mode:
	<settings> parent group containing settings.
 		<fan> contains settings related to fan control and RPM readout
			<mode> is the mode used to control the fan. Default is RPM.
				ONOFF: Only on/ off control is used
				PWM: PWM control is used, with no RPM feedback (for fans that don't have this option)
					 Manual calibration is required
				RPM: PWM control with RPM feedback is used. Calibration is performed automatically
			<ONOFFgpio> The GPIO pin for ONOFF control. Defaults to GPIO 27. This pin is used to switch on or
						off the fan in PWM and RPM mode if connected in one of these modes.
			<ONOFFinvert> Invert the ONOFF signal required for some switching hardware. Default is false.
			<PWMcalibrated> is the calibrated PWM percentage where the fan starts running. Default is 30.
							Can be calibrated by runnning smartfancontrol with the argument -c or -cal.
                            Only used in PWM mode and to measure minimum value for autocalibration.
			<PWMgpio> The GPIO pin for PWM control. Defaults to GPIO 18. Take care a hardware PWM
					  compatible pin is chosen.
			<PWMfrequency> The hardware PWM frequency in Hz. Default is 10000.
			<PWMinvert> Invert the PWM signal required for some switching hardware. Default is false.
			<recalibrate> The number of days between automatic calibrations. Default is 7. A calibration
						  will be performed n days after startup or previous calibration at 12:00 PM.
						  Only used in RPM mode.
			<RPMgpio> The GPIO pin for RPM readout. Defaults to GPIO 17. Only used in RPM mode.
			<RPMpullup> Use internal pullup for RPM GPIO pin. Default is true. Only used in RPM mode.
			<RPMppr> The number of RPM tacho pulses per revolution. Default is 2. Only used in RPM mode.
			<RPMedge> If true, both positive and negative edges are used for RPM measurement. Default is true.
                      Only used in RPM mode.
			<RPMfiltersize> If larger than 1, measured RPMs are filtered by an n-sized moving average filter
							Default is 0. Only used in RPM mode.
			<Frequency> The frequency of the fan control loop in Hz. default is 10.
			<Pgain> The P gain of the fan control loop. Default is 0.1. Only used in RPM mode.
			<Igain> The I gain of the fan control loop. Default is 0.2. Only used in RPM mode.

		<temp> contains settings related to temperature input.
			<cpu> Use CPU temperature input. Default is true.
			<hdd> Use HDD temperature input. Default is empty. Enter HDD to measure here (/dev/sdx).
			<ext> Use external temperature input. Default is empty. Enter temperature file here (/run/tempx).
			<mode> is the mode used to read the temperature. Default is MAX.
				MIN: Use the minimum temperature from all sensors used.
				AVG: Use the average temperature from all sensors used.
				MAX: Use the maximum temperature from all sensors used.
			<Farenheit> Display the temperature in Farenheit. Default is false (temperature is displayed in Celcius)
			<AlarmHigh> Temperature to rise a temperature high alarm. Default is 65 Celcius.
						If Farenheit is selected, then this temperature is in Farenheit.
			<AlarmCrit> Temperature to rise a temperature critical alarm. Default is 80 Celcius.
						If Farenheit is selected, then this temperature is in Farenheit.
			<AlarmShutdown> Will the system shutdown at critical temperature. Default is true.
							Alternatively a file/scriptname can be entered here to execute when this alarm occurs.

		<control> contains settings related to the temperature controller.
			<mode> is the mode used to control the temperature loop. Default is LINEAR.
				ONOFF: Only on/ off control is used, if this mode is selected, on/off control will also be used to
					   control the fan.
				LINEAR: If the temperature is within the linear range, the fan speed will increase linear with a
						temperature increase.
				PI: PI temperature control is used.
			<TempOn> The temperature to switch the fan on. Default is 55 Celcius. Only used in ONOFF mode.
					 If Farenheit is selected, then this temperature is in Farenheit.
			<TempHyst> The temperature hysteresis to switch the fan off again. Default is 5 Celcius.
					   Only used in ONOFF mode. If Farenheit is selected, then this temperature is in Farenheit.
			<TempStart> The temperature to start the fan controller. Default is 45 Celcius.
					    Only used in LINEAR and PI mode. If Farenheit is selected, then this temperature is in Farenheit.
			<TempFull> The temperature on which the fan controller should run full speed. Default is 65 Celcius.
					   Only used in LINEAR and PI mode. If Farenheit is selected, then this temperature is in Farenheit.
			<LinSteps> The temperature steps/ hysteresis. To provent oscillating on small temperature changes.
					   Default is 2.5 Celcius. Only used in LINEAR mode. If Farenheit is selected, then this temperature is in Farenheit.
			<Frequency> The frequency of the temperature control loop in Hz. default is 1.
			<Pgain> The P gain of the temperature control loop. Default is 10. Only used in PI mode.
			<Igain> The I gain of the temperature control loop. Default is 0.1. Only used in PI mode.

For testing and tuning the following command line parameters are available. Take care to stop the service before running commandline settings:
sudo systemctl stop smartfancontrol.service

Usage:
         smartfancontrol <args>
         -h, --help   : this help file
         -v, --version: print version information
         -c, --cal    : manually calibrate PWM level and exit
         -t, --temp   : Monitor temperature(s)
         -f, --fan    : Set fan PWM or RPM for testing
         -a, --auto   : Autotune fan PI controller (RPM control)
         -d, --dtrmn  : Determine temperature PI controller optimum parameters
         -m, --mon    : Monitor actual status in terminal
         <no argument>: run as daemon

That's all for now ...

Please send Comments and Bugreports to hellyrulez@home.nl


debian/changelog
smartfancontrol (0.86) stable; urgency=low

  * Add option to fan running in idle mode at low temperatures

 -- Helly1206 <hellyrulez@hotmail.com>  Sat, 06 Nov 2021 15:31:19 +0100

smartfancontrol (0.84) stable; urgency=low

  * Fixed small bug in alarms timer

 -- Helly1206 <hellyrulez@hotmail.com>  Sat, 25 Jul 2021 10:51:37 +0100

smartfancontrol (0.83) stable; urgency=low

  * Added leading zero when monitor temperature is displayed in Farenheit

 -- Helly1206 <hellyrulez@hotmail.com>  Sat, 02 Jan 2021 19:12:17 +0100

smartfancontrol (0.82) stable; urgency=low

  * Removed absolute dependency to pigpio as it is not installable in ubuntu

 -- Helly1206 <hellyrulez@hotmail.com>  Fri, 01 Jan 2021 09:12:17 +0100

smartfancontrol (0.81) stable; urgency=low

  * Small changes to package builder

 -- Helly1206 <hellyrulez@hotmail.com>  Mon, 27 Apr 2020 19:58:53 +0100

smartfancontrol (0.80) stable; urgency=low

  * Initial release

 -- Helly1206 <hellyrulez@hotmail.com>  Sat, 25 Apr 2020 8:57:23 +0100


opt/smartfancontrol/hardware/temp.py
# -*- coding: utf-8 -*-
#########################################################
# SERVICE : temp.py                                     #
#           Measures the temperature from:              #
#           * CPU                                       #
#           * HDD                                       #
#           * external                                  #
#           (or function, average or maximum)           #
#           I. Helwegen 2020                            #
#########################################################

####################### IMPORTS #########################
from common.common import common
from subprocess import Popen, PIPE
#########################################################

####################### GLOBALS #########################
MODE_MIN = -1
MODE_AVG = 0
MODE_MAX = 1

ALARM_NONE = 0
ALARM_HIGH = 10
ALARM_CRIT = 100

CPU_TEMP = "/sys/class/thermal/thermal_zone0/temp"
HDD_TEMP = ["smartctl", "-a"]
DEF_EXT_LOC = "/run/temp"
DEF_SHUTDOWN = ["shutdown", "-h", "now"]

MONITOR_SLEEP = 1

ABS_NULL = -273.15

#########################################################

###################### FUNCTIONS ########################

#########################################################

#########################################################
# Class : temp                                          #
#########################################################

"""
<temp> contains settings related to temperature input.
<cpu> Use CPU temperature input. Default is true.
<hdd> Use HDD temperature input. Default is empty. Enter HDD to measure here (/dev/sdx).
<ext> Use external temperature input. Default is empty. Enter temperature file here (/run/tempx).
<mode> is the mode used to read the temperature. Default is MAX. 
    MIN: Use the minimum temperature from all sensors used. 
    AVG: Use the average temperature from all sensors used.
    MAX: Use the maximum temperature from all sensors used.
<Farenheit> Display the temperature in Farenheit. Default is false (temperature is displayed in Celcius)
<AlarmHigh> Temperature to rise a temperature high alarm. Default is 65 Celcius. 
            If Farenheit is selected, then this temperature is in Farenheit.
<AlarmCrit> Temperature to rise a temperature critical alarm. Default is 80 Celcius. 
            If Farenheit is selected, then this temperature is in Farenheit.
<AlarmShutdown> Will the system shutdown at critical temperature. Default is true.
                Alternatively a file/scriptname can be entered here to execute when this alarm occurs.
"""

class temp(common):
    def __init__(self, settings, alarm, logger = None):
        self.alarm = alarm
        self.logger = logger
        common.__init__(self, self.logger)
        
        self.cpu = self.checkkey(settings,'temp','cpu')
        self.hdd = self.checkkey(settings,'temp','hdd')
        self.ext = self.checkkey(settings,'temp','ext')
        if self.ext:
            if type(self.ext) != str:
                self.ext = DEF_EXT_LOC

        mode = self.checkkey(settings,'temp','mode')
        if mode.lower() == 'min':
            self.mode = MODE_MIN
        elif mode.lower() == 'max':
            self.mode = MODE_MAX
        else:
            self.mode = MODE_AVG
            
        self.Farenheit = self.checkkey(settings,'temp','Farenheit')
        self.AlarmHigh = self.checkkey(settings,'temp','AlarmHigh')
        self.AlarmCrit = self.checkkey(settings,'temp','AlarmCrit')
        AlarmShutdown = self.checkkey(settings,'temp','AlarmShutdown')
        if AlarmShutdown:
            if type(AlarmShutdown) == str:
                self.AlarmShutdown = AlarmShutdown.split()
            else:
                self.AlarmShutdown = DEF_SHUTDOWN
        else:
            self.AlarmShutdown = None
        self.temperature = None
        self.curalarm = ALARM_NONE

    def __del__(self):
        pass
    
    def __str__(self):
        return self.print(self.temperature)
    
    def __repr__(self):
        tempstr = "-"
        if self.temperature:
            if self.Farenheit:
                tempstr = "0{:.1f}".format(self.Celcius2Farenheit(self.temperature))
            else: 
                tempstr = "{:.1f}".format(self.temperature)
        return tempstr
    
    def print(self, temp):
        if not temp:
            if self.Farenheit:
                tempstr = "-'F"
            else:
                tempstr = "-'C"
        else:
            if self.Farenheit:
                tempstr = "{:.1f}'F".format(self.Celcius2Farenheit(temp))
            else:
                 tempstr = "{:.1f}'C".format(temp)
        return tempstr
    
    def update(self):
        Temp = []
        tCPU = self.GetCPUTemp()
        if tCPU:
            Temp.append(tCPU)
        tHDD = self.GetHDDTemp()
        if tHDD:
            Temp.append(tHDD)
        tEXT = self.GetEXTTemp()
        if tEXT:
            Temp.append(tEXT)
        if len(Temp) > 0:
            if self.mode == MODE_MIN:
                self.temperature = min(Temp)   
            elif self.mode == MODE_MAX:
                self.temperature = max(Temp)
            else:
                self.temperature = sum(Temp)/len(Temp)
        else:
            self.temperature = None
        
        return self.temperature, self.getalarm(), tCPU, tHDD, tEXT
    
    def get(self):
        if self.temperature:
            return self.temperature
        else:
            return ABS_NULL
    
    def monitor(self, exitevent):
        print("Monitoring temperature")
        while not exitevent.is_set():
            vals = self.update()
            print("{} [CPU: {}, HDD: {}, EXT: {}] {}".format(self.print(vals[0]), self.print(vals[2]), self.print(vals[3]), self.print(vals[4]), repr(self.alarm)))
            exitevent.wait(MONITOR_SLEEP)
        print("Finished monitoring temperature")
        return
    
    def getalarm(self):
        if self.temperature == None:
            if not self.alarm.get(self.alarm.ALARM_TEMPNONE):
                self.alarm.set(self.alarm.ALARM_TEMPNONE)
                self.alarm.reset(self.alarm.ALARM_TEMPHIGH)
                self.alarm.reset(self.alarm.ALARM_TEMPCRIT)
                self.loge(self.alarm)
            return self.alarm.ALARM_TEMPNONE
        elif self.temperature >= self.AlarmCrit:
            if not self.alarm.get(self.alarm.ALARM_TEMPCRIT):
                self.alarm.set(self.alarm.ALARM_TEMPCRIT)
                self.loge(self.alarm)
                if self.AlarmShutdown:
                    self.loge("Shutting down system")
                    try:
                        Popen(self.AlarmShutdown)
                    except:
                        self.loge("Shutdown failed, system may overheat !!!!!")
                else:
                    self.loge("No shutdown intended, system may overheat !!!!!")
            return self.alarm.ALARM_TEMPCRIT
        elif self.temperature >= self.AlarmHigh:
            if not self.alarm.get(self.alarm.ALARM_TEMPHIGH):
                self.alarm.set(self.alarm.ALARM_TEMPHIGH)
                self.alarm.reset(self.alarm.ALARM_TEMPCRIT)
                self.logw(self.alarm)
            return self.alarm.ALARM_TEMPHIGH
        else:
            self.alarm.reset(self.alarm.ALARM_TEMPNONE)
            self.alarm.reset(self.alarm.ALARM_TEMPHIGH)
            self.alarm.reset(self.alarm.ALARM_TEMPCRIT)
            return self.alarm.ALARM_NONE
    
    def GetCPUTemp(self):
        if self.cpu:
            try:
                with open(CPU_TEMP,"r") as f:
                    CPUTemp = float(f.read())/1000.0
                return CPUTemp
            except:
                return None
        else:
            return None    
        
    def GetHDDTemp(self):
        #sudo smartctl -A /dev/sda | grep Temperature_Celsius | awk '{print $10}'
        ####194 Temperature_Celsius     0x0022   117   107   000    Old_age   Always       -       30
        if self.hdd:
            try:
                HDDTemp = None
                for line in Popen(HDD_TEMP + [str(self.hdd)], stdout=PIPE).stdout.read().decode("utf-8").split('\n'):
                    if ('Temperature_Celsius' in line.split()) or ('Temperature_Internal' in line.split()) or ('Airflow_Temperature_Cel' in line.split()):
                        HDDTemp = float(line.split()[9])
                return HDDTemp
            except:
                return None
        else:
            return None
    
    def GetEXTTemp(self):
        if self.ext:
            try:
                with open(self.ext,"r") as f:
                    EXTTemp = float(f.read())
                return EXTTemp
            except:
                return None
        else:
            return None
        
    def Celcius2Farenheit(self, tempc):
        return tempc * 1.8 + 32

######################### MAIN ##########################
if __name__ == "__main__":
    pass
